File: content/problems/0792-number-of-matching-subsequences/solutions/waiting-character-buckets/Solution.java
class Solution {
    public int numMatchingSubseq(String s,String[] words){
        List<Deque<int[]>> buckets=new ArrayList<>();
        for(int i=0;i<26;i++)buckets.add(new ArrayDeque<>());
        for(int i=0;i<words.length;i++)buckets.get(words[i].charAt(0)-'a').addLast(new int[]{
            i,0
        });
        int answer=0;
        for(char c:s.toCharArray()){
            Deque<int[]> queue=buckets.get(c-'a');
            int size=queue.size();
            while(size-->0){
                int[] state=queue.removeFirst();
                state[1]++;
                if(state[1]==words[state[0]].length())answer++;
                else buckets.get(words[state[0]].charAt(state[1])-'a').addLast(state);
            }
        }
        return answer;
    }
}

File: content/problems/0792-number-of-matching-subsequences/solutions/waiting-character-buckets/Solution.kt
class Solution {
    fun numMatchingSubseq(s:String,words:Array<String>):Int{
        val buckets=Array(26){
            java.util.ArrayDeque<IntArray>()
        };
        for(i in words.indices)buckets[words[i][0]-'a'].addLast(intArrayOf(i,0));
        var answer=0;
        for(c in s){
            val queue=buckets[c-'a'];
            repeat(queue.size){
                val state=queue.removeFirst();
                state[1]++;
                if(state[1]==words[state[0]].length)answer++ else buckets[words[state[0]][state[1]]-'a'].addLast(state)
            }
        };
        return answer
    }
}
